### PR TITLE
Adds Reserved Contributors to Multi-Contributor Contract

### PR DESCRIPTION
**Pull request by darcys22**
At the deployment of the multi contributor contract the service node operator has the ability to specify a number of reserved contributors that must contribute to the contract for it to become a service node.

### CODE REVIEW

File: contracts/ServiceNodeContribution.sol [hunk 1]
```diff
@@ -404,7 +484,7 @@ contract ServiceNodeContribution is Shared {
      */
     function minimumContribution() public view returns (uint256 result) {
         result = calcMinimumContribution(
-            stakingRequirement - totalContribution(),
+            stakingRequirement - totalContribution() - totalReservedContribution(),
             contributorAddresses.length,
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:488]: This `numContributors` parameter value here needs to include unfilled, but reserved addresses as well.

File: contracts/ServiceNodeContribution.sol [hunk 2]
```diff
@@ -86,6 +91,11 @@ contract ServiceNodeContribution is Shared {
         operator = tx.origin; // NOTE: Creation is delegated by operator through factory
         blsPubkey = _blsPubkey;
         serviceNodeParams = _serviceNodeParams;
+
+        for (uint256 i = 0; i < _reservedContributors.length; i++) {
+            reservedContributorAddresses.push(_reservedContributors[i].addr);
+            reservedContributions[_reservedContributors[i].addr] += _reservedContributors[i].stakedAmount;
+        }
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:98]: This needs to test as it goes that the reserved contribution amounts are meeting the minimum contribution requirement, given the operator stake + any previous reserved contributions added during the loop.    That is, `[25% operator, 10%, 10%, 15%, 50%]` and `[25% operator, 70%, 5%]` are acceptable reservations, but `[25% operator, 10%, 5%]` is not.

But that said, see my other comment about the operator contribution not being known at this point.

(There are also some slightly weird cases where `[25% operator, 70%, 5%]` is an acceptable reservation, but `[25%, 5%, 70%]` is not, but we have those limitations in current OXEN reservations as well, so I don't think it's necessary to worry about them in the contract; the front-end/contract caller can just sort them in descending order first if it wants to allow those sorts of things.)

File: contracts/ServiceNodeContribution.sol [hunk 3]
```diff
@@ -86,6 +91,11 @@ contract ServiceNodeContribution is Shared {
         operator = tx.origin; // NOTE: Creation is delegated by operator through factory
         blsPubkey = _blsPubkey;
         serviceNodeParams = _serviceNodeParams;
+
+        for (uint256 i = 0; i < _reservedContributors.length; i++) {
+            reservedContributorAddresses.push(_reservedContributors[i].addr);
+            reservedContributions[_reservedContributors[i].addr] += _reservedContributors[i].stakedAmount;
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:97]: Needs a check that this won't exceed the staking requirement.  (But also see comment about operator contribution not being known yet).

File: contracts/ServiceNodeContribution.sol [hunk 4]
```diff
@@ -76,7 +80,8 @@ contract ServiceNodeContribution is Shared {
         address _stakingRewardsContract,
         uint256 _maxContributors,
         BN256G1.G1Point memory _blsPubkey,
-        IServiceNodeRewards.ServiceNodeParams memory _serviceNodeParams
+        IServiceNodeRewards.ServiceNodeParams memory _serviceNodeParams,
+        IServiceNodeRewards.Contributor[] memory _reservedContributors
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:84]: There are two problems with having this in the constructor:
- it isn't updateable if the operator wants to reuse the contract
- acceptable reserved contributions *depend* on the amount the operator contributes (both for "does this meet the minimum" and "does this not exceed the staking requirement" checks), and the operator contribution isn't known here.

I think both can be solved by taking `_reservedContributors` as a parameter of `contributeOperatorFunds` (or perhaps having an alternative version called something like `contributeOperatorFundsWithReservations`), since that seems to be the earliest point where we know the operator amount and thus can validate the reservation amounts.

**Comment by darcys22** [comment]
@jagerman review notes addressed in #d2b7f5d

File: contracts/ServiceNodeContribution.sol [hunk 5]
```diff
@@ -111,12 +115,31 @@ contract ServiceNodeContribution is Shared {
      */
     function contributeOperatorFunds(
         uint256 amount,
-        IServiceNodeRewards.BLSSignatureParams memory _blsSignature
+        IServiceNodeRewards.BLSSignatureParams memory _blsSignature,
+        IServiceNodeRewards.Contributor[] memory _reservedContributors
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:119]: Does solidity allow a `= []` default value for this argument so that you can just call it without the reserved arguments?

File: contracts/ServiceNodeContribution.sol [hunk 6]
```diff
@@ -225,7 +263,8 @@ contract ServiceNodeContribution is Shared {
      * be calculated by calling `calcMinimumContribution` with the staking
      * requirement and 0 contributors.
      */
-    function resetContract(uint256 amount) external onlyOperator {
+    function resetContract(uint256 amount, IServiceNodeRewards.Contributor[] memory reservedContributors) external onlyOperator {
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:266]: Perhaps `= []` default on reservedContributors, if that's something solidity allows.

File: contracts/ServiceNodeContribution.sol [hunk 7]
```diff
@@ -225,7 +263,8 @@ contract ServiceNodeContribution is Shared {
      * be calculated by calling `calcMinimumContribution` with the staking
      * requirement and 0 contributors.
      */
```
Comment (jagerman) [inline contracts/ServiceNodeContribution.sol:265]:
```suggestion
     * @param _reservedContributors array of address/amounts to reserve minimum stakes
     * for the given addresses.  Can be empty to leave contributor spots open to anyone.
     */
```

File: contracts/ServiceNodeContribution.sol (hunk 5 above)
Comment (darcys22) [reply contracts/ServiceNodeContribution.sol:119]: AFAIK it doesnt - they have a discussion here which implies it is not planned either https://github.com/ethereum/solidity/issues/232

File: contracts/ServiceNodeContribution.sol (hunk 6 above)
Comment (darcys22) [reply contracts/ServiceNodeContribution.sol:266]: See comment on other. But not available in solidity

File: contracts/ServiceNodeContribution.sol (hunk 5 above)
Comment (jagerman) [reply contracts/ServiceNodeContribution.sol:121]: No worries then.  It does support overloading, so I suppose we could have a second version with the same name without reserved contributions, but it's probably not worthwhile versus just requiring an empty array here.
